### PR TITLE
ddl: Add `CreatePlacementPolicyWithInfo` to `ddl.DDL`

### LINKED ISSUE #32951: Add `CreatePlacementPolicyWithInfo` to interface `ddl.DDL` to support br restore
required by #32290 to help br restore placement policies

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #32951

Problem Summary:

`CreatePlacementPolicyWithInfo` is required by `BR` to restore placement policies

### What is changed and how it works?

1. Add `CreatePlacementPolicyWithInfo` to `ddl.DDL`
2. To support `or replace` for  `CreatePlacementPolicyWithInfo` , move the replace logical to `onCreatePlacementPolicy`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

File: ddl/ddl.go [hunk 1]
```diff
@@ -147,6 +147,12 @@ type DDL interface {
 		info []*model.TableInfo,
 		onExist OnExist) error
 
+	// CreatePlacementPolicyWithInfo creates a placement policy
+	//
+	// WARNING: the DDL owns the `policy` after calling this function, and will modify its fields
```
Comment (bb7133) [inline ddl/ddl.go:152]: > WARNING: the DDL owns the `policy` after calling this function

What does 'own' mean exactly? Is it something like the 'ownership' in Rust?

Comment (lcwangchao) [reply ddl/ddl.go:152]: This comment is copied from `CreateTableWithInfo` .  But yes, I think it is something like 'ownership' in Rust, the policy in the argument will be referred by the job in args. I'm not sure what will happens if we use it after calling the function, but it's better not to do it.

File: ddl/placement_policy.go [hunk 2]
```diff
@@ -29,21 +29,47 @@ import (
...
-	err = checkPolicyValidation(policyInfo.PlacementSettings)
+
+	existPolicy, err := getPlacementPolicyByName(d, t, policyInfo.Name)
 	if err != nil {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+
+	if existPolicy != nil {
+		if !orReplace {
```
Comment (3pointer) [inline ddl/placement_policy.go:51]: is there any chance OnExistIgnore policy ddl reach here 🤔？

Comment (bb7133) [reply ddl/placement_policy.go:51]: Good point, it is possible when a placement policy(say `S`) does not exist before a DDL job is submitted( https://github.com/pingcap/tidb/pull/32986/files#diff-b7c54f7dad7257cbe4fa04b8d76c9bed694e2e2e7c5599f45eb2172d8d4ff8a7R2329) but being created when the DDL job is in the queue and executed here.

It happens when tow `CREATE PLACEMENT POLICY IF NOT EXISTS ...` executed concurrently in multiple clients. However, reporting an error might be a better result for the users, although it seems weird.

**Pull request merged**